### ISSUE TITLE
Issue 1 : DMP Compass Output incorrect

### ISSUE THREAD

**Issue by dobodu** [dobodu/ICM20948_DMP_Micropython#1]
Data received directly from ICM20948

Accel: 00.04   00.03    01.03
Gyro:  00.21 -0.17 00.11
Mag:   -26.10 -8.40 -97.65

From DMP

Accel:          0.0314	 0.0314   1.0039  => Acceleration OK
Gyroscope  0.0000	  0.0610    0.0000  ==> It's also OK (single spot value + DLPF)
Mag            - 1.0000  0.0000	514.0000  ==> THERE MUST BE A BUG SOMEWHERE ?

**Comment by dobodu** [comment]
Still working on it.

Here comes an exemple of decoded FIFO Outputs for Raw magnetometer (value changes when I rotate the device)
Very VERY annoying not to find the reason

DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz -1.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz -1.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz -1.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz -1.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz -1.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz -1.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz -1.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz -1.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz -1.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz -1.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz -1.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 0.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 0.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 0.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 0.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 0.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 0.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 257.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 257.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 257.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 257.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 257.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 257.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -1.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my -1.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 257.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 257.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 257.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 257.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 257.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 257.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 257.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 257.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx 0.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my 0.0000	mz 514.0000

**Comment by dobodu** [comment]
Well, since version 0.9-6.5, the bud is still there

The Slave register reads correctly data from AK-9916 RSV2 register and seem to feed DMP correctly but ouput is still the same...

**Comment by JeromeRaymond** [comment]
I have made tests 
If I execute te code bellow, I see a result 
DBG:	 ICM20948 :	 FIFO Quaternion_9	q1 0.0032	q2 -0.0067	q3 -0.0000	accuracy 257.0000
q1, q2 and q3 seems to be yaw pitch and roll informations in radian. Values are not compensated by magnetic detector but it seems work

The code
from machine import Pin, I2C
import icm20948
from math import degrees, atan2
from utime import sleep_ms
from ustruct import unpack_from


i2c0 = I2C(0, scl=Pin(9), sda=Pin(8), freq=400000)


#dmp = False for direct Acc/Gyr/Magn
#dmp = True for DMP processor

imu = icm20948.ICM20948(i2c0, dmp = True, debug=True)

if not imu.dmp_ready :
    
    imu.gyro_cal()
    imu.mag_cal()
    #imu.DMP_enable_sensor()
    
    while True:
        a = imu.acc()
        g = imu.gyro()
        m = imu.mag()   # Read directly to the slave
        print(m)
        #imu.Q_update_nomag(a, g)
        #imu.Q_update_full(a, g, m)
        heading = imu.heading
        pitch = imu.pitch
        roll = imu.roll

        print(f"""
Accel: {a[0]:05.2f} {a[1]:05.2f} {a[2]:05.2f}
Gyro:  {g[0]:05.2f} {g[1]:05.2f} {g[2]:05.2f}
Mag:   {m[0]:05.2f} {m[1]:05.2f} {m[2]:05.2f}
HPT:   {heading:7.3f} {pitch:7.3f} {roll:7.3f}""")
        
        sleep_ms(25)
    
else :
           
    #imu.DMP_enable_sensor("ACCELEROMETER",True)
    #imu.DMP_enable_sensor("GYROSCOPE",True)
    #imu.DMP_enable_sensor("RAW_ACCELEROMETER",True)
    #imu.DMP_enable_sensor("RAW_GYROSCOPE",True)
    #imu.DMP_enable_sensor("MAGNETIC_FIELD_UNCALIBRATED",True)
    #imu.DMP_enable_sensor("GYROSCOPE_UNCALIBRATED",True)
    #imu.DMP_enable_sensor("ACTIVITY_CLASSIFICATON",True)
    #imu.DMP_enable_sensor("STEP_DETECTOR",True)
    #imu.DMP_enable_sensor("STEP_COUNTER",True)
    #imu.DMP_enable_sensor("GAME_ROTATION_VECTOR",True)
    #imu.DMP_enable_sensor("ROTATION_VECTOR",True)
    #imu.DMP_enable_sensor("GEOM_ROTATION_VECTOR",True)
    #imu.DMP_enable_sensor("GEOM_FIELD",True)
    #imu.DMP_enable_sensor("GRAVITY",True)
    #imu.DMP_enable_sensor("LINEAR_ACCELERATION",True)
    imu.DMP_enable_sensor("ORIENTATION",True)
    
    while True:
        imu.DMP_fifo_read()

**Comment by dobodu** [comment]
> I have made tests 
> If I execute te code bellow, I see a result 
> DBG:	 ICM20948 :	 FIFO Quaternion_9	q1 0.0032	q2 -0.0067	q3 -0.0000	accuracy 257.0000
> q1, q2 and q3 seems to be yaw pitch and roll informations in radian. Values are not compensated by magnetic detector but it seems work


Thank you for the feedback. 

I think all the DMP reports works fine But the magnetometer. Thus the lack of magnetic compensation is a normal issue for now. 

I think I missed something in the way I set up ak9916 or I transmis wrong data's to the DMP. 
I'll post my bug tracking reflections xion and question by the end of the way. 

Don't hesitate to ask question related to the library code. For sure I don't see something.

**Comment by dobodu** [comment]
So I'll try to explain more precisely the problem I have. Step by step following the library

When initialising the library we do :

`self.read(0, ICM_WHO_AM_I) == ICM_CHIP_ID`

To check the ICM20948 chip : **No Problem**, then 

```
self.reg_config(0,ICM_PWR_MGMT_1,
ICM_PWR_MGMT_1_RESET, True)
```

We reset the chip : **no problem**, then 

`self.write(0, ICM_PWR_MGMT_1, ICM_PWR_MGMT_1_CLOCK_AUTO)`

We setup Bank 0, PWR_MGNT_1 to Clock_Auto : **No problem**, then 

`self.write(0, ICM_PWR_MGMT_2, 0x00)`

We activate all sensor : **No problem**, then 

`self.write(3,` ICM_I2C_MST_CTRL, ICM_I2C_MST_CTRL_NSR | 0x07)  #I2C MSTR CLOCK = 07 = 345,6kHz`

We set-up ICM_I2C_MST_CTRL for clock 0x07 = 345,6 kHz and ask for a stop between reads : **No problems** then

`self.reg_config(0, ICM_USER_CTRL, ICM_USER_CTRL_I2C_MST_EN, True)`

We Enable I2C Master Control : **No problems** then 

```
self.slave_config(0, AK_I2C_ADDR, AK_WIA2, 1, True, True, False, False, False)
if self.read(0, ICM_EXT_SLV_SENS_DATA_00) != AK_CHIP_ID:
      self._dbg("AK09916 Magnetometer Chip Not Found... ")
```

We, ask slave 0 to read AK09916 Who am I, (1st True = Read, 2nd True = Enable)
and then check if we can read the correct Who Am I byte : **No problem** then 

```
self.slave_config(0, AK_I2C_ADDR, AK_CNTL3, 1, False, True, False, False, False,AK_CNTL3_RESET)
self.slave_config(0, AK_I2C_ADDR, AK_CNTL3, 1, True, True, False, False, False)
while self.read(0, ICM_EXT_SLV_SENS_DATA_00) == 0x01: #Loop util reset bit remains on
     sleep_ms(10)
self._dbg("AK09916 Magnetometer Chip Reseted... ")
```

Here we fist  write (1st False) AK_CNTL3_RESET bit to AK_CNTL3 
Then ask for reading AK_CNTL3 to check the reset status **No Problem**

Up to this **NO PROBLEMS ARE ENCOUTERED**

**Comment by dobodu** [comment]
Then we have Two options, 

First One is using ICM20948 without DMP (dmp=False) : this is done by ICM_Config and consist in :

```
self.slave_config(0, AK_I2C_ADDR, AK_CNTL2, 1 , False, True, False, False, False, AK_CNTL2_MODE_100HZ)
self.slave_config(0, AK_I2C_ADDR, AK_ST1, 9, True, True, False, False, False)
```

Then we can read directly Compass data correctly by accessing Slave 0 : **This is working perfectly !**

**Comment by dobodu** [comment]
Second option is using DMP (dmp = True) : this is done by DMP_Config and consist in : 

```
self.reg_config(0,ICM_PWR_MGMT_1, ICM_PWR_MGMT_1_SLEEP, False)
self.reg_config(0, ICM_PWR_MGMT_1, ICM_PWR_MGMT_1_LP, False)
```

We ask ICM to exit SLEEP or LOW POWER MODE : **No problem** then 
```

self.slave_config(0, AK_I2C_ADDR, AK_RSV2, 10, True, True, True, False, True)
self.slave_config(1, AK_I2C_ADDR, AK_CNTL2, 1,   False, True, False, False, False, AK_CNTL2_MODE_SINGLE)
```

We ask Slave 0 to read 10 bytes (True = Read, True = Enable, True = Swap bytes, and last True = Grouped bytes)
and we telle Slave 1 to write A byte SINGLE_MODE to AK_CNTL2

I consider **no problem** since reading directly Slave 0 shows the compass date (Big indian etc...,),
for example I can read x -6.3 y -22.2 z 82.5 uT starting at (0x05 and not 0x04 as said in Sparkfun comments but has no impact
since we read both 10 bytes starting at AK_RSV2 adress = 0x03)

Let's keep on, next we do 

`self.write(3, ICM_I2C_MST_ODR_CONFIG, 0x04)`

We ask I2C MASTER to OUTPUT DATE at 1000/2^4 = 68,75 Hz exactly as Sparkun library
```
self.reg_config(0, ICM_LP_CFG,  ICM_LP_CFG_MST, True)
self.reg_config(0, ICM_LP_CFG,  ICM_LP_CFG_ACC | ICM_LP_CFG_GYRO , False)
```

We tell that ACC and GYRO are not in LP Mode but I2C_MASTER is in LP Mode (as expected in Sparkfun Lib)

`self.reg_config(0,ICM_USER_CTRL, ICM_USER_CTRL_DMP_EN | ICM_USER_CTRL_FIFO_EN , False)`

We disable DMP and FIFO, and then setup Gyro Full Scale and Acc Full scale (I assume it has no relation to compasse value so let's swipe that part), then

```
self.write(0, ICM_FIFO_EN_1, 0x00)
self.write(0, ICM_FIFO_EN_2, 0x00)
self.write(0, ICM_FIFO_RST,0x1F)
self.write(0, ICM_FIFO_RST,0x1E)
```

We turn off FIFO (as expected in Sparkfun) and reset the FIFO (1F thant 1E)
from there we set up Acc and gyro Sample rate (as below, no relation to compass so let's swipe this part)
and finally we reach the DMP firmware loading 

```
self.DMP_load_firmware()
self._buffer = bytearray(2)
self._buffer[0] = DMP_START_ADDRESS >> 8
self._buffer[1] = DMP_START_ADDRESS & 0xff
self.write_bytes(2, ICM_PRGM_START_ADDRH, self._buffer)
```

This is writting (correctly I assume since DMP outputs responds correctly tho sensor activation and output what was required)
the DMP firmware to the internal chip, and tells that the DMP start adress (DMP_START_ADDRESS = 0x1000)

**I believe everything is correct till this point.**  but if anyone see something incorrect please help

**Comment by JeromeRaymond** [comment]
For compass calibration, you should look to this other projet.
(https://headtracker.gitbook.io/head-tracker-v2.2/getting-started/imu-calibration-and-drift)

**Comment by dobodu** [comment]
Finally we setup different things : 

```
self.write(0, ICM_HW_FIX_DISABLE,0x48)
self.write(0, ICM_SINGLE_FIFO_PRIORITY_SEL,0xE4)
```

We write 0x48 to ICM HW FIX DISABLE and 0xE4 to SINGLE FIFO PRIORITY (as expected in sparkfun library)

```
DMP_ACC_SCALE_FACTOR = [0x04, 0x00, 0x00, 0x00]
self.DMP_write(DMP_ACC_SCALE, DMP_ACC_SCALE_FACTOR)
DMP_ACC_SCALE_FACTOR2 = [0x00, 0x04, 0x00, 0x00]
self.DMP_write(DMP_ACC_SCALE2, DMP_ACC_SCALE_FACTOR2)
```

We write to DMP ACC SCALE FACTORS (as Sparkfun lib)

```
DMP_COMPAS_MOUNT_MATRIX_SCALED_ZERO = [0x00, 0x00, 0x00, 0x00]
DMP_COMPAS_MOUNT_MATRIX_SCALED_PLUS1 = [0x09, 0x99, 0x99, 0x99]
DMP_COMPAS_MOUNT_MATRIX_SCALED_MINUS1 = [0xF6, 0x66, 0x66, 0x67]
self.DMP_write(DMP_CPASS_MTX_00, DMP_COMPAS_MOUNT_MATRIX_SCALED_PLUS1)
self.DMP_write(DMP_CPASS_MTX_01, DMP_COMPAS_MOUNT_MATRIX_SCALED_ZERO)
self.DMP_write(DMP_CPASS_MTX_02, DMP_COMPAS_MOUNT_MATRIX_SCALED_ZERO)
self.DMP_write(DMP_CPASS_MTX_10, DMP_COMPAS_MOUNT_MATRIX_SCALED_ZERO)
self.DMP_write(DMP_CPASS_MTX_11, DMP_COMPAS_MOUNT_MATRIX_SCALED_MINUS1)
self.DMP_write(DMP_CPASS_MTX_12, DMP_COMPAS_MOUNT_MATRIX_SCALED_ZERO)
self.DMP_write(DMP_CPASS_MTX_20, DMP_COMPAS_MOUNT_MATRIX_SCALED_ZERO)
self.DMP_write(DMP_CPASS_MTX_21, DMP_COMPAS_MOUNT_MATRIX_SCALED_ZERO)
self.DMP_write(DMP_CPASS_MTX_22, DMP_COMPAS_MOUNT_MATRIX_SCALED_MINUS1)
```

We setup compass mounting matrix

```
DMP_B2S_MOUNT_MATRIX_SCALED_ZERO = [0x00, 0x00, 0x00, 0x00]
DMP_B2S_MOUNT_MATRIX_SCALED_PLUS1 = [0x40, 0x00, 0x00, 0x00]
self.DMP_write(DMP_CPASS_MTX_00, DMP_B2S_MOUNT_MATRIX_SCALED_PLUS1)
self.DMP_write(DMP_CPASS_MTX_01, DMP_B2S_MOUNT_MATRIX_SCALED_ZERO)
etc...
```

we setup B2S Mounting matrix then we have a setup related to  Gyro_SF, Gyro Scale, Acc Only Gain and so on...
that to my belief has nothing to do with compass issue since DMP accelerometer values are OK.

then 

```
DMP_CPASS_TIME_BUFFER_FACTOR = [0x00, 0x45]
self.DMP_write(DMP_CPASS_TIME_BUFFER, DMP_CPASS_TIME_BUFFER_FACTOR)
```

We setup DMP Compass to Ox45 = 69 Hz (as in Sparkfun library) and 

**Up to there we closely follow Sparkfun library story....**

**Comment by dobodu** [comment]
Finally we do : 

```
self.reg_config(0, ICM_USER_CTRL, ICM_USER_CTRL_DMP_EN | ICM_USER_CTRL_FIFO_EN, True)
self.reg_config(0, ICM_USER_CTRL, ICM_USER_CTRL_DMP_RST, True)
self.reg_config(0,ICM_PWR_MGMT_1, ICM_PWR_MGMT_1_LP, True)
```

We enable DMP and FIFO, then reset DMP (this line has no impact on the problem, if suppressed the behaviour si still the same) and put the chip in LP mode....

From there, I have

DBG:	 ICM20948 :	 	Header is 0x2000
DBG:	 ICM20948 :	 FIFO Compass		mx -1.0000	my -1.0000	mz 514.0000	Time 1877586
DBG:	 ICM20948 :	 FIFO Compass		mx -258.0000	my -258.0000	mz 257.0000	Time 2067502 (changing orientation for exemple)

mx, my, mz values are changing, but always by +/- 256 steps **(VERY STRANGE ???)**

My interpretation is that the compass Slave reading is correct, feeding correctly the DMP, but I must do something wrong...

**Comment by dobodu** [comment]
> For compass calibration, you should look to this other projet. (https://headtracker.gitbook.io/head-tracker-v2.2/getting-started/imu-calibration-and-drift)

Thanks Jerome, but it's definitly not a calibration issue, for now, the purpose of the code is that DMP processor must output exactly the same values as direct reading to AK9916, no matter calibration is done or not...
But it might be usefull for the following work to do after I manage to solve this sticky bug....

**Comment by JeromeRaymond** [comment]
for your message "mx, my, mz values are changing, but always by +/- 256 steps (VERY STRANGE ???)" maybe values are read with a wrong beginning (8 bytes on the right)
I have downloaded the latest icm20948.py file but I can't see compass information when DMP is OFF.
Can you add a test script to get all datas ?
Thank you

**Comment by JeromeRaymond** [comment]
in your last modified file, Row 832, I have an error about an unknown value about accmin. Other values are magmin

**Comment by dobodu** [comment]
### I FOUND IT

I FOUND IT !

The error was in no way related to DMP or compass mount matrix.

Let(s explain my mystage (for Pyhon User)

```
from struct import unpack_from

DMP_Pedom_Quat6_Byte_Ordering = [1, 0, 3, 2, 5, 4]
DMP_Compass_Bytes = 6

print()
print("First Method Error")
print("==================")

data = bytearray(b'\xff\x92\xff\xf1\x02\x7f')
data_ordered = data     <==== The mistake is here
print("Raw data", data)
for i in range(DMP_Compass_Bytes) :
    data_ordered[DMP_Pedom_Quat6_Byte_Ordering[i]] = data[i]
print("Ordered data", data_ordered)
mx,my,mz = unpack_from("<3h", data_ordered)
print(mx, my, mz)

print()
print("Second Method Success")
print("=====================")

data = bytearray(b'\xff\x92\xff\xf1\x02\x7f')
data_ordered = bytearray(DMP_Compass_Bytes)      <==== No mote mistake, new object
print("Raw data", data)
for i in range(DMP_Compass_Bytes) :
    data_ordered[DMP_Pedom_Quat6_Byte_Ordering[i]] = data[i]
print("Ordered data", data_ordered)
mx,my,mz = unpack_from("<3h", data_ordered)
print(mx, my, mz)
```

First Method Error
============
Raw data bytearray(b'\xff\x92\xff\xf1\x02\x7f')
Ordered data bytearray(b'\xff\xff\xff\xff\x02\x02')
-1 -1 514

Second Method 
===========
Raw data bytearray(b'\xff\x92\xff\xf1\x02\x7f')
Ordered data bytearray(b'\x92\xff\xf1\xff\x7f\x02')
-110 -15 639

In other words
==========

I was working on a memory copy of the byterray ==> Values transfered from original byterray were applyed to itself making garbage....

I'm so sad, I made this basic error !

**Comment by dobodu** [comment]
Bug found and library corrected
